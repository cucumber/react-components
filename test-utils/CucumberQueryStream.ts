import * as messages from '@cucumber/messages'
import { Query } from '@cucumber/query'
import { Writable } from 'stream'

export class CucumberQueryStream extends Writable {
  constructor(private readonly query: Query) {
    super({ objectMode: true })
  }

  _write(envelope: messages.Envelope, _: string, callback: (error?: Error | null) => void): void {
    this.query.update(envelope)
    callback(null)
  }
}
