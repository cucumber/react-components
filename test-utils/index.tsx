import { Query as GherkinQuery } from '@cucumber/gherkin-utils'
import { Query as CucumberQuery } from '@cucumber/query'
import { render } from '@testing-library/react'
import React, { FunctionComponent, PropsWithChildren, ReactElement } from 'react'

import CucumberQueryContext from '../src/CucumberQueryContext.js'
import GherkinQueryContext from '../src/GherkinQueryContext.js'
import UriContext from '../src/UriContext.js'

export interface TestRenderOptions {
  uri?: string
  gherkinQuery?: GherkinQuery
  cucumberQuery?: CucumberQuery
}

const AllTheProviders: FunctionComponent<PropsWithChildren<{ options: TestRenderOptions }>> = ({
  children,
  options,
}) => {
  return (
    <GherkinQueryContext.Provider value={options.gherkinQuery ?? new GherkinQuery()}>
      <UriContext.Provider value={options.uri ?? 'some.feature'}>
        <CucumberQueryContext.Provider value={options.cucumberQuery ?? new CucumberQuery()}>
          {children}
        </CucumberQueryContext.Provider>
      </UriContext.Provider>
    </GherkinQueryContext.Provider>
  )
}

const customRender = (ui: ReactElement, options: Partial<TestRenderOptions> = {}) => {
  const WrappedWithOptions: FunctionComponent<PropsWithChildren> = ({ children }) => (
    <AllTheProviders options={options}>{children}</AllTheProviders>
  )
  return render(ui, { wrapper: WrappedWithOptions })
}

export * from './CucumberQueryStream.js'
export * from './runFeature.js'
export * from './search.js'
export * from '@testing-library/react'

export { customRender as render }
