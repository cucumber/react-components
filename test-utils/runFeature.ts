import { IHook, ISupportCodeExecutor, runCucumber, SupportCode } from '@cucumber/fake-cucumber'
import { makeSourceEnvelope } from '@cucumber/gherkin'
import { GherkinStreams } from '@cucumber/gherkin-streams'
import { Query as GherkinQuery } from '@cucumber/gherkin-utils'
import * as messages from '@cucumber/messages'
import { Writable } from 'stream'

export class FailingCodeSupport implements ISupportCodeExecutor {
  constructor(readonly stepDefinitionId: string) {}

  public execute() {
    throw new Error('Woops ...')
  }
  public argsToMessages(): messages.StepMatchArgument[] {
    return []
  }
}

export class FailingHook implements IHook {
  constructor(public readonly id: string) {}

  match(): ISupportCodeExecutor {
    return new FailingCodeSupport('')
  }

  toMessage(): messages.Envelope {
    return {
      hook: {
        id: this.id,
        sourceReference: {
          uri: __dirname,
        },
      },
    }
  }
}

export async function runFeature(
  feature: string,
  gherkinQuery: GherkinQuery,
  supportCode: SupportCode = new SupportCode()
): Promise<messages.Envelope[]> {
  const emitted: messages.Envelope[] = []
  const out = new Writable({
    objectMode: true,
    write(
      envelope: messages.Envelope,
      encoding: string,
      callback: (error?: Error | null) => void
    ): void {
      emitted.push(envelope)
      try {
        callback()
      } catch (err) {
        callback(err)
      }
    },
  })

  const gherkinEnvelopeStream = GherkinStreams.fromSources(
    [makeSourceEnvelope(feature, 'test.feature')],
    {
      newId: messages.IdGenerator.incrementing(),
    }
  )

  await runCucumber(supportCode, gherkinEnvelopeStream, gherkinQuery, out)
  return emitted
}
