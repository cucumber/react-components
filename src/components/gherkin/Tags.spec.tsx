import React from 'react'

import { render } from '../../../test-utils'
import { Tags } from './Tags'

describe('Tags', () => {
  it('doesnt render anything if no tags', () => {
    const { container } = render(<Tags tags={[]} />)
    expect(container).toBeEmptyDOMElement()
  })

  it('renders if we really have some tags', () => {
    const tags = [
      {
        location: {
          line: 1,
        },
        name: '@foo',
        id: '1',
      },
      {
        location: {
          line: 3,
        },
        name: '@bar',
        id: '2',
      },
    ]
    const { asFragment } = render(<Tags tags={tags} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
