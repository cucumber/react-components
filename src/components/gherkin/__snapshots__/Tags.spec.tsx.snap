// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags renders if we really have some tags 1`] = `
<DocumentFragment>
  <ul>
    <li>
      <span
        class="highlight"
      >
        @foo
      </span>
    </li>
    <li>
      <span
        class="highlight"
      >
        @bar
      </span>
    </li>
  </ul>
</DocumentFragment>
`;
